### PR TITLE
Enable raspberry 3B to use camera, add raspberry-camera documnet

### PR DESCRIPTION
**Pull request by gfwm2013**
使树莓派3B可以调用摄像头进行预测，并加上树莓派3B摄像头的安装与激活文档

### CODE REVIEW

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+sudo raspistill -o test.jpg 
+```
+
+之后使用树莓派自带的`xdg-open`打开图片
+
+```shell
+sudo xdg-open test.jpg
+```
+
+## 5、opencv可以使用树莓派的摄像头
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/Raspberry-camera.md]: 5、安装V4L2驱动使OpenCV能够识别摄像头

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+## 1、树莓派Raspberry Pi camera的购买
+
+树莓派Raspberry Pi camera可以从淘宝、京东等平台购买。
+
+## 2、安装Raspberry Pi camera
+
+将树莓派 3B 断电后，将摄像头如下图安装在树莓派上。
+
+![](../doc/raspberry_with_camera.png)
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+执行命令
+
+```shell
+sudo raspi-config
+```
+进入设置页面后，之后按照下面步骤设置
+
+1. 选择 `Interfacing Options`
+
+![](../doc/raspberry_enable_cameta_step1.png)
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 4]
```diff
@@ -112,7 +114,9 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
                                 cv::Mat &output_image) {
   const int TOPK = 3;
   int max_indices[TOPK];
+  std::memset(max_indices, 0, sizeof(max_indices));
   double max_scores[TOPK];
+  std::memset(max_scores, 0.0, sizeof(max_scores));
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: Done.

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 5]
```diff
@@ -135,41 +139,61 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
-    printf("Usage: \nimage_classification_demo model_dir label_path "
-           "input_image_path "
-           "output_image_path\n");
-    return -1;
-  }
-  std::string model_dir = argv[1];
-  std::string label_path = argv[2];
-  std::string input_image_path = argv[3];
-  std::string output_image_path = argv[4];
+cv::Mat video_detect(cv::Mat& input_image, std::vector<std::string> word_labels,
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
+# ARMLinux（树莓派3B）摄像头的购买、安装、使能与验证
+
+## 1、树莓派Raspberry Pi camera的购买
+
+树莓派Raspberry Pi camera可以从淘宝、京东等平台购买。
+
+## 2、安装Raspberry Pi camera
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+将树莓派 3B 断电后，将摄像头如下图安装在树莓派上。
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+![](../doc/raspberry_with_camera.png)
+
+## 3、树莓派 3B 使能摄像头
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+![](../doc/raspberry_enable_cameta_step3.png)
+
+4. 点击 `Ok`
+
+![](../doc/raspberry_enable_cameta_step4.png)
+
+5. 之后重启树莓派
+
+## 4、测试摄像头是否使能成功
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md (hunk 1 above)
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 10]
```diff
@@ -135,97 +140,158 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
-  config.set_power_mode(CPU_POWER_MODE);
-
-  // 2. Create PaddlePredictor by MobileConfig
-  std::shared_ptr<paddle::lite_api::PaddlePredictor> predictor =
-      paddle::lite_api::CreatePaddlePredictor<paddle::lite_api::MobileConfig>(
-          config);
-
-  // 3. Preprocess image and fill the data of input tensor
+cv::Mat detection(bool enable_camera, cv::Mat &input_image,
+                  std::vector<std::string> word_labels,
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: 使用引用const std::vector<std::string> &word_labels,

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 11]
```diff
@@ -135,97 +140,158 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
-  // imshow("image classification demo", output_image);
-  printf("results: %d\n", results.size());
-  for (int i = 0; i < results.size(); i++) {
-    printf("Top%d %s - %f\n", i, results[i].class_name.c_str(),
-           results[i].score);
+  double postprocess_end_time = get_current_us();
+  double postprocess_time = (postprocess_end_time - postprocess_start_time) / 1000.0f;
+  printf("Postprocess time: %f ms\n\n", postprocess_time);
+
+  if (!enable_camera) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: 没有必要区分

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 12]
```diff
@@ -135,97 +140,158 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
-  double total_time_cost = 0.0f;
-  for (int i = 0; i < REPEAT_COUNT; i++) {
-    auto start = get_current_us();
-    predictor->Run();
-    auto end = get_current_us();
-    double cur_time_cost = (end - start) / 1000.0f;
-    if (cur_time_cost > max_time_cost) {
-      max_time_cost = cur_time_cost;
+  double prediction_time;
+  if (!enable_camera) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: 没有必要区分，在main函数里判断并修改WARMUP_COUNT和REAPT_COUNT数就行了

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 13]
```diff
@@ -23,8 +23,8 @@
 #include <fstream>
 
 const int WARMUP_COUNT = 1;
-const int REPEAT_COUNT = 1;
-const int CPU_THREAD_NUM = 1;
+const int REPEAT_COUNT = 5;
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: 改成可修改的吧int WARMUP_COUNT = 1; int REPEAT_COUNT = 5;

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 14]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+             results[i].bottom);
+    }
+  }
+  return output_image;
+}
+
+int main(int argc, char **argv) {
+  if (argc < 3 || argc == 4) {
+    printf(
+        "Usage: \nIf you want to use video to do pretict, please set"
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 这个修改下 Usage: \n object_detection_demo model_dir label_path [input_image_path] [output_image_path]\n
disable fetching images from camera if input_image_path and input_image_path is set.

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 15]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+      paddle::lite_api::CreatePaddlePredictor<paddle::lite_api::MobileConfig>(config);
+
+  if (enable_camera) {
+    cv::VideoCapture cap(-1);
+    cap.set(CV_CAP_PROP_FRAME_WIDTH, 640);
+    cap.set(CV_CAP_PROP_FRAME_HEIGHT, 480);
+    if (!cap.isOpened()) {
+      return -1;
+    }
+
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 去掉回车

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 16]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+    return -1;
+  }
+
+  std::string model_dir = argv[1];
+  std::string label_path = argv[2];
+  std::string input_image_path;
+  std::string output_image_path;
+  bool enable_camera = true;
+
+  if (argc > 3) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 去掉这部分代码

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 17]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+
+  paddle::lite_api::MobileConfig config;
+  config.set_model_dir(model_dir);
+  config.set_threads(CPU_THREAD_NUM);
+  config.set_power_mode(CPU_POWER_MODE);
+
+  std::shared_ptr<paddle::lite_api::PaddlePredictor> predictor =
+      paddle::lite_api::CreatePaddlePredictor<paddle::lite_api::MobileConfig>(config);
+
+  if (enable_camera) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: if (argc > 3) {
  // enable camera is input&output image path is not set
  WARMUP_COUNT = 0;
  REPEAT_COUNT = 1;
} else {
  std::string input_image_path = argv[3];
  std::string output_image_path = argv[4];
  cv::Mat input_image = cv::imread(input_image_path, 1);
  cv::Mat output_image = detection(enable_camera, input_image, word_labels, predictor);
  cv::imwrite( output_image_path, output_image);
  cv::imshow("object detection demo", output_image);
  cv::waitKey(0);
}

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 18]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+        "\n\tobject_detection_demo model_dir label_path\n"
+        "\nIf you want to use image to do pretict, please set"
+        "\n\tobject_detection_demo model_dir label_path input_image_path "
+        "output_image_path\n");
+    return -1;
+  }
+
+  std::string model_dir = argv[1];
+  std::string label_path = argv[2];
+  std::string input_image_path;
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 去掉input_image_path和output_image_path定义

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 19]
```diff
@@ -144,100 +146,164 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+    cap.set(CV_CAP_PROP_FRAME_HEIGHT, 480);
+    if (!cap.isOpened()) {
+      return -1;
+    }
+
+    while (1) {
+      cv::Mat input_image;
+      cap >> input_image;
+      cv::Mat output_image = detection(enable_camera, input_image, word_labels, predictor);
+      cv::imshow("Object Detection", output_image);
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 和cv::imshow("object detection demo", output_image);保持一致

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 20]
```diff
@@ -0,0 +1,62 @@
+# ARMLinux（树莓派3B）摄像头的购买、安装、使能与验证
+
+## 1、树莓派Raspberry Pi camera的购买
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+树莓派Raspberry Pi camera可以从淘宝、京东等平台购买。
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/Raspberry-camera.md [hunk 22]
```diff
@@ -0,0 +1,62 @@
+# ARMLinux（树莓派3B）摄像头的购买、安装、使能与验证
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/Raspberry-camera.md]: Done.

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 23]
```diff
@@ -220,24 +247,90 @@ int main(int argc, char **argv) {
...
+    if (!cap.isOpened()){
+      return -1;
+    }
+
+    while(1){
+      cv::Mat frame;	  
+      cap >> frame;
+      cv::Mat input_image = frame.clone();
+      cv::Mat output_image = video_detect(input_image, word_labels, predictor);
+      cv::imshow("Predictor CAM", output_image);
```
[the remark this answers is not in the document]
Comment (gfwm2013) [reply PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: Done.

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 24]
```diff
@@ -144,100 +146,146 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
+    REPEAT_COUNT = 5;
+    std::string input_image_path = argv[3];
+    std::string output_image_path = argv[4];
+    cv::Mat input_image = cv::imread(input_image_path);
+    cv::Mat output_image = process(input_image, word_labels, predictor);
+    cv::imwrite(output_image_path, output_image);
+    cv::imshow("Object Detection Demo", output_image);
+    cv::waitKey(0);
+  }
+  else{
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 做下代码格式化
} else {

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 25]
```diff
@@ -144,100 +146,146 @@ std::vector<RESULT> postprocess(const float *output_data, int64_t output_size,
...
     output_size *= dim;
+  }
   cv::Mat output_image = input_image.clone();
+  double postprocess_start_time = get_current_us();
   std::vector<RESULT> results = postprocess(
-      output_data, output_size, word_labels, SCORE_THRESHOLD, output_image);
-  cv::imwrite(output_image_path, output_image);
-  // imshow("object detection demo", output_image);
+      output_data, output_size, word_labels, SCORE_THRESHOLD, output_image, prediction_time);
+  double postprocess_end_time = get_current_us();
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 做下代码格式化

File: PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc [hunk 26]
```diff
@@ -203,29 +186,93 @@ int main(int argc, char **argv) {
...
   }
-  cv::waitKey(0);
+  printf("Preprocess time: %f ms\n", preprocess_time);
+  printf("Prediction time: %f ms\n", prediction_time);
+  printf("Postprocess time: %f ms\n\n", postprocess_time);
+  return output_image;
+}
+
+int main(int argc, char **argv) {
+  if (argc < 3 || argc == 4) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/image_classification_demo/image_classification_demo.cc]: 4应该改成5吧

File: PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc [hunk 27]
```diff
@@ -214,30 +195,91 @@ int main(int argc, char **argv) {
...
-  cv::waitKey(0);
+  printf("Preprocess time: %f ms\n", preprocess_time);
+  printf("Prediction time: %f ms\n", prediction_time);
+  printf("Postprocess time: %f ms\n\n", postprocess_time);
+
+  return output_image;
+}
+
+int main(int argc, char **argv) {
+  if (argc < 3 || argc == 4) {
```
Comment (hong19860320) [inline PaddleLite-armlinux-demo/object_detection_demo/object_detection_demo.cc]: 4改成5

**Pull request merged**